tests/integration/suite/httpserver/httpserver.go

```diff
@@ -0,0 +1,159 @@
+	h1Client := http.Client{Transport: &http.Transport{}}
+	h2cClient := &http.Client{
+		Transport: &http2.Transport{
+			// Allow http2.Transport to use protocol "http"
+			AllowHTTP: true,
+			// Pretend we are dialing a TLS endpoint
+			DialTLS: func(network, addr string, cfg *tls.Config) (net.Conn, error) {
+				return net.Dial(network, addr)
+			},
+		},
+	}
+
+	// Wait for daprd to be up
+	assert.Eventually(t, func() bool {
+		reqCtx, reqCancel := context.WithTimeout(ctx, 150*time.Millisecond)
+		defer reqCancel()
+		req, err := http.NewRequestWithContext(reqCtx, http.MethodGet, fmt.Sprintf("http://localhost:%d/v1.0/healthz", h.proc.HTTPPort()), nil)
+		if err != nil {
+			return false
+		}
+
+		// Body is closed below but the linter isn't seeing that
+		//nolint:bodyclose
+		res, err := h1Client.Do(req)
+		if err != nil {
+			return false
+		}
+
+		err = closeBody(res.Body)
+		if err != nil {
+			return false
+		}
+
+		return res.StatusCode == http.StatusNoContent
+	}, time.Second*5, 200*time.Millisecond)
+
+	t.Run("test with HTTP1", func(t *testing.T) {
+		reqCtx, reqCancel := context.WithTimeout(ctx, 150*time.Millisecond)
+		defer reqCancel()
+
+		req, err := http.NewRequestWithContext(reqCtx, http.MethodGet, fmt.Sprintf("http://localhost:%d/v1.0/healthz", h.proc.HTTPPort()), nil)
+		require.NoError(t, err)
+
+		// Body is closed below but the linter isn't seeing that
+		//nolint:bodyclose
+		res, err := h1Client.Do(req)
+		require.NoError(t, err)
+		defer closeBody(res.Body)
+		require.Equal(t, http.StatusNoContent, res.StatusCode)
+
+		assert.Equal(t, 1, res.ProtoMajor)
+	})
+
+	t.Run("test with HTTP2 Cleartext with prior knowledge", func(t *testing.T) {
+		reqCtx, reqCancel := context.WithTimeout(ctx, 150*time.Millisecond)
+		defer reqCancel()
+
+		req, err := http.NewRequestWithContext(reqCtx, http.MethodGet, fmt.Sprintf("http://localhost:%d/v1.0/healthz", h.proc.HTTPPort()), nil)
+		require.NoError(t, err)
+
+		// Body is closed below but the linter isn't seeing that
+		//nolint:bodyclose
+		res, err := h2cClient.Do(req)
+		require.NoError(t, err)
+		defer closeBody(res.Body)
+		require.Equal(t, http.StatusNoContent, res.StatusCode)
+
+		assert.Equal(t, 2, res.ProtoMajor)
+	})
+
+	t.Run("server allows upgrading from HTTP1 to HTTP2 Cleartext", func(t *testing.T) {
+		// The Go HTTP client doesn't handle upgrades from HTTP/1 to HTTP/2
+		// The best we can do for now is to verify that the server responds with a 101 response when we signal that we are able to upgrade
+		// See: https://github.com/golang/go/issues/46249
+
+		reqCtx, reqCancel := context.WithTimeout(ctx, 150*time.Millisecond)
+		defer reqCancel()
+
+		req, err := http.NewRequestWithContext(reqCtx, http.MethodGet, fmt.Sprintf("http://localhost:%d/v1.0/healthz", h.proc.HTTPPort()), nil)
+		require.NoError(t, err)
+
+		req.Header.Set("Connection", "Upgrade, HTTP2-Settings")
+		req.Header.Set("Upgrade", "h2c")
+		req.Header.Set("HTTP2-Settings", "AAMAAABkAARAAAAAAAIAAAAA")
```

Comment:
any specific magic string here?

Reply:
That is from the specs: https://book.hacktricks.xyz/pentesting-web/h2c-smuggling

Users will never need to worry about this stuff. HTTP clients should be smart to do these things automatically. In this case, we are doing things manually to confirm stuff works as expected.